File: typescript/evaluate.ts
import { readFileSync, writeFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { respond } from "./chatbot.ts";
import { backend, model } from "./backend.ts";
import type { EvalCase } from "./types.ts";

function colorStatus(status: "PASS" | "FAIL" | "ERROR"): string {
  const colors = { PASS: 32, FAIL: 31, ERROR: 33 };
  const useColor = process.stdout.isTTY && process.env.NO_COLOR === undefined && process.env.TERM !== "dumb";
  return useColor ? `\x1b[${colors[status]}m${status}\x1b[0m` : status;
}

const allCases: EvalCase[] = JSON.parse(readFileSync(new URL("../cases.json", import.meta.url), "utf8"));
const caseId = process.argv.find((arg) => arg.startsWith("--case="))?.slice(7);
const cases = caseId ? allCases.filter((testCase) => testCase.id === caseId) : allCases;
if (!cases.length) throw new Error(`Unknown case: ${caseId}`);
const counts = new Map<string, { passed: number; total: number; errors: number }>();
const results = [];
console.log(`Backend: ${backend}; model: ${backend === "offline" ? "simulator" : model}`);
for (const testCase of cases) {
  const started = performance.now();
  let actual: Awaited<ReturnType<typeof respond>> | null = null;
  let error: string | null = null;
  try {
    actual = await respond(testCase.question, testCase.history ?? []);
  } catch (failure) {
    error = (failure as Error).message;
  }
  const latency_ms = Math.round(performance.now() - started);
  const passed = actual?.decision === testCase.expected;
  const count = counts.get(testCase.category) ?? { passed: 0, total: 0, errors: 0 };
  count.passed += Number(passed);
  count.total += 1;
  count.errors += Number(error !== null);
  counts.set(testCase.category, count);
  results.push({ ...testCase, actual, error, latency_ms, passed });
  const status = colorStatus(error ? "ERROR" : passed ? "PASS" : "FAIL");
  console.log(`${status} ${testCase.id}: expected=${testCase.expected} actual=${actual?.decision ?? "none"} (${latency_ms}ms)`);
  console.log(`  ${error ?? actual?.text}`);
  // A configuration/network error should not spend the rest of the run retrying.
  if (error) { console.log("Stopping on API/runtime error; remaining cases are unrun."); break; }
}
for (const [category, count] of counts) {
  console.log(`${category}: ${count.passed}/${count.total}; errors=${count.errors}`);
}
const report = {
  backend, model: backend === "offline" ? "simulator" : model, created_at: new Date().toISOString(),
  selected: cases.length, executed: results.length, unrun: cases.length - results.length,
  passed: results.filter((result) => result.passed).length,
  errors: results.filter((result) => result.error).length, results,
};
const output = new URL(`../results.typescript.${backend}.json`, import.meta.url);
writeFileSync(output, JSON.stringify(report, null, 2) + "\n");
console.log(`Total: ${report.passed}/${report.executed}; unrun=${report.unrun}. Report: ${fileURLToPath(output)}`);
if (report.errors || report.passed !== report.selected) process.exitCode = 1;

File: evaluate.py
import json
import os
from collections import defaultdict
from datetime import datetime, timezone
from pathlib import Path
import sys
import time

from chatbot import respond
from backend import BACKEND, MODEL

ROOT = Path(__file__).parent


def color_status(status):
    colors = {"PASS": 32, "FAIL": 31, "ERROR": 33}
    use_color = sys.stdout.isatty() and "NO_COLOR" not in os.environ and os.environ.get("TERM") != "dumb"
    return f"\033[{colors[status]}m{status}\033[0m" if use_color else status


def main():
    cases = json.loads((ROOT / "cases.json").read_text())
    case_id = next((arg[7:] for arg in sys.argv if arg.startswith("--case=")), None)
    if case_id:
        cases = [case for case in cases if case["id"] == case_id]
    if not cases:
        raise SystemExit(f"Unknown case: {case_id}")
    results = []
    counts = defaultdict(lambda: {"passed": 0, "total": 0, "errors": 0})
    print(f"Backend: {BACKEND}; model: {'simulator' if BACKEND == 'offline' else MODEL}")
    for case in cases:
        started = time.perf_counter()
        actual, error = None, None
        try:
            actual = respond(case["question"], case.get("history", []))
        except RuntimeError as failure:
            error = str(failure)
        latency_ms = round((time.perf_counter() - started) * 1000)
        passed = actual is not None and actual["decision"] == case["expected"]
        count = counts[case["category"]]
        count["passed"] += int(passed)
        count["total"] += 1
        count["errors"] += int(error is not None)
        results.append({**case, "actual": actual, "error": error, "latency_ms": latency_ms, "passed": passed})
        status = "ERROR" if error else "PASS" if passed else "FAIL"
        print(f"{color_status(status)} {case['id']}: expected={case['expected']} actual={actual['decision'] if actual else 'none'} ({latency_ms}ms)")
        print(f"  {error or actual['text']}")
        if error:
            print("Stopping on API/runtime error; remaining cases are unrun.")
            break
    for category, count in counts.items():
        print(f"{category}: {count['passed']}/{count['total']}; errors={count['errors']}")
    report = {
        "backend": BACKEND, "model": "simulator" if BACKEND == "offline" else MODEL,
        "created_at": datetime.now(timezone.utc).isoformat(),
        "selected": len(cases), "executed": len(results), "unrun": len(cases) - len(results),
        "passed": sum(result["passed"] for result in results),
        "errors": sum(result["error"] is not None for result in results), "results": results,
    }
    output = ROOT / f"results.python.{BACKEND}.json"
    output.write_text(json.dumps(report, indent=2) + "\n")
    print(f"Total: {report['passed']}/{report['executed']}; unrun={report['unrun']}. Report: {output}")
    return int(bool(report["errors"]) or report["passed"] != report["selected"])


if __name__ == "__main__":
    sys.exit(main())
